### PR TITLE
Backend auth

### LINKED ISSUE #1: Backend Auth
Set up auth backend using Passport.js & Google OAuth
Examples:
- [LAH](https://github.com/hack4impact-uiuc/life-after-hate/blob/master/backend/utils/passport-setup.js)
- [Recruitment Portal](https://github.com/hack4impact-uiuc/h4i-recruitment/blob/master/backend/src/middleware/passport.js)

### PR DESCRIPTION
**Pull request by Qrtn**
## Summary

Adds support for authentication via Google OAuth.

## Changes

- Backend endpoints for authentication
- Sets up Passport
- Adds cookie (client-side) session support to keep track of user being authenticated
- Make most database fields not required so we can create the user and fill in information later
- Add middleware for protecting requests to certain levels of authentication

## Requests / Responses 

**Request**
GET `/api/auth/user` If user is logged in, returns info, otherwise returns null

**Response**
HTTP/1.1 200 OK

```
{
  "code": 200,
  "data": USER_INFO,
  "success": true
}
```

**Request**
GET `/api/auth/login` Redirects the user to the Google login page. If authentication is successful, a cookie is set on the client-side. This cookie allows the client to access authenticated routes. Finally, redirects to `/`.

**Request**
POST `/api/auth/logout` Clears the cookie so that the client can no longer access authenticated routes.

**Response**
HTTP/1.1 200 OK

```
{
  "code": 200,
  "message": "Logged out.",
  "success": true
}
```

Closes #1

### CODE REVIEW

File: api/src/api/auth.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
+const express = require('express');
+const router = express.Router();
+const passport = require('passport');
+
+router.get('/user', (req, res) => {
+  const user = req.user || null;
```
Comment (mattwalo32) [inline api/src/api/auth.js:6]: Isn't this just the same as ```const user = req.user```

File: api/src/api/home.js [hunk 2]
```diff
@@ -16,4 +17,17 @@ router.get(
   }),
 );
 
+router.get(
+  '/registeredOnly',
```
Comment (mattwalo32) [inline api/src/api/home.js:21]: Is this endpoint only for testing?

File: api/src/passport-setup.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    {
+      clientID: process.env.GOOGLE_CLIENT_ID,
+      clientSecret: process.env.GOOGLE_CLIENT_SECRET,
+      callbackURL: process.env.OAUTH_CALLBACK_URI,
+    },
+    async (accessToken, refreshToken, profile, cb) => {
+      // find the user in the database based on their oauth id
+      const user = await Member.findOne({ oauthID: profile.id });
+
+      console.log(profile);
```
Comment (mattwalo32) [inline api/src/passport-setup.js:34]: Remove log message

File: api/test/dummy.test.js [hunk 4]
```diff
@@ -0,0 +1,3 @@
+describe('Dummy test', () => {
+  test('Dummy test', () => {});
```
Comment (mattwalo32) [inline api/test/dummy.test.js:2]: why the dummy test

**Review by mattwalo32 — CHANGES_REQUESTED** [review]
Nice, very clean (especially like your auth.js file). Just a few changes

File: api/src/api/home.js (hunk 2 above)
Comment (Qrtn) [reply api/src/api/home.js:21]: Yes, left this endpoint in there in order to show how to use one of the auth middlewares. We'll remove all of home.js eventually.

File: api/test/dummy.test.js (hunk 4 above)
Comment (Qrtn) [reply api/test/dummy.test.js]: Added it in there to make the tests pass for deployment but I saw you fixed this with `yarn test --passWithNoTests`, thanks!

**Review by mattwalo32 — APPROVED** [review]

**Review by amit-sawhney — APPROVED** [review]

**Pull request merged**